File: Homework/Ngan_Luu/20-valid-parentheses.py
https://leetcode.com/problems/valid-parentheses/
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        for char in s:
            if char in ['(','{','[']:
                stack.append(char)
            else:
                if len(stack)>0:
                    if stack[-1]+char in ['{}','()','[]']:
                        stack.pop()
                    else:
                        return False
                else:
                    return False
        if len(stack)==0:
            return True

File: Homework/Son/W2/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
 var isValid = function(s) {
    if (s.length === 1) return false
    let parenthesisStack = []
    for (let c of s) {
        if (c === "(" || c === "{" || c === "[") parenthesisStack.push(c)
        switch (c) {
            case ")":
                if (parenthesisStack.length === 0) return false;
                if (parenthesisStack[parenthesisStack.length -1] === "(") parenthesisStack.pop();
                else return false;
                break;
            case "}":
                if (parenthesisStack.length === 0) return false;
                if (parenthesisStack[parenthesisStack.length -1] === "{") parenthesisStack.pop();
                else return false;
                break;
            case "]":
                if (parenthesisStack.length === 0) return false;
                if (parenthesisStack[parenthesisStack.length -1] === "[") parenthesisStack.pop();
                else return false;
                break;
            default:
                break;
        }
    }
    if (parenthesisStack.length > 0) return false
    else return true
};

// https://leetcode.com/problems/valid-parentheses/